File: pki.go
package main

import (
    "crypto"
    "crypto/ecdsa"
    "crypto/elliptic"
    "crypto/rand"
    "crypto/rsa"
    "crypto/sha256"
    "crypto/x509"
    "encoding/asn1"
    "encoding/base64"
    "encoding/gob"
    "fmt"
    "io"
    "math/big"
    "os"
)

type PKI struct {
    signingPriv *ecdsa.PrivateKey
    encryptingPriv *rsa.PrivateKey
}

func (p *PKI) Init() error {
    file := "pki_keys.gob"
    if _, err := os.Stat(file); err == nil {
        f, err := os.Open(file)
        if err != nil {
            return err
        }
        defer f.Close()
        dec := gob.NewDecoder(f)
        var keys struct {
            SigningPriv []byte
            EncryptingPriv []byte
        }
        if err := dec.Decode(&keys); err != nil {
            return err
        }
        signingPrivBlock, _ := pem.Decode(keys.SigningPriv)
        p.signingPriv, err = x509.ParseECPrivateKey(signingPrivBlock.Bytes)
        if err != nil {
            return err
        }
        encryptingPrivBlock, _ := pem.Decode(keys.EncryptingPriv)
        p.encryptingPriv, err = x509.ParsePKCS1PrivateKey(encryptingPrivBlock.Bytes)
        if err != nil {
            return err
        }
        return nil
    }

    // Generate keys
    var err error
    p.signingPriv, err = ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
    if err != nil {
        return err
    }
    p.encryptingPriv, err = rsa.GenerateKey(rand.Reader, 2048)
    if err != nil {
        return err
    }

    // Encode to PEM
    signingPrivPEM := pem.EncodeToMemory(&pem.Block{Type: "EC PRIVATE KEY", Bytes: x509.MarshalECPrivateKey(p.signingPriv)})
    encryptingPrivPEM := pem.EncodeToMemory(&pem.Block{Type: "RSA PRIVATE KEY", Bytes: x509.MarshalPKCS1PrivateKey(p.encryptingPriv)})

    keys := struct {
        SigningPriv    []byte
        EncryptingPriv []byte
    }{
        SigningPriv:    signingPrivPEM,
        EncryptingPriv: encryptingPrivPEM,
    }

    f, err := os.Create(file)
    if err != nil {
        return err
    }
    defer f.Close()
    enc := gob.NewEncoder(f)
    if err := enc.Encode(keys); err != nil {
        return err
    }
    return nil
}

func (p *PKI) Sign(message string) (string, error) {
    if p.signingPriv == nil {
        return "", fmt.Errorf("keys not initialized")
    }
    h := sha256.Sum256([]byte(message))
    r, s, err := ecdsa.Sign(rand.Reader, p.signingPriv, h[:])
    if err != nil {
        return "", err
    }
    type ecdsaSignature struct {
        R, S *big.Int
    }
    sigBytes, err := asn1.Marshal(ecdsaSignature{R: r, S: s})
    if err != nil {
        return "", err
    }
    return base64.StdEncoding.EncodeToString(sigBytes), nil
}

func (p *PKI) Verify(message, signatureB64 string) (bool, error) {
    if p.signingPriv == nil {
        return false, fmt.Errorf("keys not initialized")
    }
    sigBytes, err := base64.StdEncoding.DecodeString(signatureB64)
    if err != nil {
        return false, err
    }
    h := sha256.Sum256([]byte(message))
    type ecdsaSignature struct {
        R, S *big.Int
    }
    var sig ecdsaSignature
    if _, err := asn1.Unmarshal(sigBytes, &sig); err != nil {
        return false, err
    }
    return ecdsa.Verify(asn1SignatureToPublicKey(p.signingPriv.PublicKey), h[:], sig.R, sig.S), nil
}

func asn1SignatureToPublicKey(pub ecdsa.PublicKey) *ecdsa.PublicKey {
    return &pub
}

func (p *PKI) Encrypt(message string) (string, error) {
    if p.encryptingPriv == nil {
        return "", fmt.Errorf("keys not initialized")
    }
    ciphertext, err := rsa.EncryptOAEP(sha256.New(), rand.Reader, &p.encryptingPriv.PublicKey, []byte(message), nil)
    if err != nil {
        return "", err
    }
    return base64.StdEncoding.EncodeToString(ciphertext), nil
}

func (p *PKI) Decrypt(ciphertextB64 string) (string, error) {
    if p.encryptingPriv == nil {
        return "", fmt.Errorf("keys not initialized")
    }
    ciphertext, err := base64.StdEncoding.DecodeString(ciphertextB64)
    if err != nil {
        return "", err
    }
    plaintext, err := rsa.DecryptOAEP(sha256.New(), rand.Reader, p.encryptingPriv, ciphertext, nil)
    if err != nil {
        return "", err
    }
    return string(plaintext), nil
}

func main() {
    var pki PKI
    if err := pki.Init(); err != nil {
        fmt.Println("Error:", err)
        return
    }
    sig, _ := pki.Sign("Hello World")
    fmt.Println("Signature:", sig)
    verified, _ := pki.Verify("Hello World", sig)
    fmt.Println("Verify:", verified)
    enc, _ := pki.Encrypt("Secret message")
    fmt.Println("Encrypted:", enc)
    dec, _ := pki.Decrypt(enc)
    fmt.Println("Decrypted:", dec)
}

File: pki.ml
open Stdlib
open Mirage_crypto
open Mirage_crypto_ec
open Mirage_crypto_pk.Rsa
open Cstruct
open Digestif.SHA256
open Base64

let file = "pki_keys.bin"

type keys = {
  signing_priv_cs : string;  (* serialized EC priv */
  signing_pub_cs : string;   (* serialized EC pub */
  rsa_p : string;            (* Z.to_cstruct p */
  rsa_q : string;            (* Z.to_cstruct q */
  rsa_e : string;            (* Z.to_cstruct e */
}

let rng = Mirage_crypto_rng.create ()

let z_of_string s = Cstruct.of_string s |> Z.of_cstruct_be
let string_of_z z = Z.to_cstruct z |> Cstruct.to_string

let load_keys () =
  let ic = open_in_bin file in
  let keys = Marshal.from_channel ic in
  close_in ic;
  keys

let save_keys keys =
  let oc = open_out_bin file in
  Marshal.to_channel oc keys [];
  close_out oc

let init () =
  try
    ignore (load_keys ());
    ()
  with End_of_file ->
    (* Generate signing keypair (ECDSA P-256) *)
    let signing_priv = generate rng `P256 in
    let sk = match signing_priv with `P256 sk -> sk in
    let signing_priv_cs = P256.priv_to_cstruct sk |> to_string in
    let signing_pub = pub_of_priv signing_priv in
    let pk = match signing_pub with `P256 pk -> pk in
    let signing_pub_cs = P256.point_to_cstruct pk |> to_string in

    (* Generate encrypting keypair (RSA 2048) *)
    let p, q = Mirage_crypto_pk.gen_prime ~bits:1024 rng, Mirage_crypto_pk.gen_prime ~bits:1024 rng in
    let e = Z.of_int 65537 in
    let encrypting_priv = priv_of_primes ~e p q in
    let rsa_p = string_of_z p in
    let rsa_q = string_of_z q in
    let rsa_e = string_of_z e in

    let keys = { signing_priv_cs; signing_pub_cs; rsa_p; rsa_q; rsa_e } in
    save_keys keys

let get_keys () =
  let k = load_keys () in
  (* Reconstruct signing keys *)
  let signing_priv_cs = of_string k.signing_priv_cs in
  let sk = P256.cstruct_to_priv signing_priv_cs |> function Ok sk -> sk | Error _ -> failwith "invalid priv" in
  let signing_priv = `P256 sk in
  let signing_pub_cs = of_string k.signing_pub_cs in
  let pk = P256.cstruct_to_point signing_pub_cs |> function Ok pk -> pk | Error _ -> failwith "invalid pub" in
  let signing_pub = `P256 pk in

  (* Reconstruct encrypting keys *)
  let p = z_of_string k.rsa_p in
  let q = z_of_string k.rsa_q in
  let e = z_of_string k.rsa_e in
  let encrypting_priv = priv_of_primes ~e p q in
  let encrypting_pub = pub_of_priv encrypting_priv in

  signing_priv, signing_pub, encrypting_priv, encrypting_pub

let sign message =
  let signing_priv, _, _, _ = get_keys () in
  let digest = digest_string (Bytes.of_string message) |> to_raw_string |> of_string in
  let k_bits = 256 in
  let k = Z.(random ~rng ~bits:k_bits) in
  let (r, s) = P256.Dsa.sign ~key:signing_priv ~k digest |> function Ok x -> x | Error _ -> failwith "sign failed" in
  let sig_cs = P256.Dsa.sig_of_cstruct (r, s) |> function Ok cs -> cs | Error _ -> failwith "sig serialize failed" in  (* assume sig_of_cstruct *)
  encode_string (to_string sig_cs)

let verify message signature_b64 =
  let _, signing_pub, _, _ = get_keys () in
  let sig_bytes = decode_string signature_b64 in
  let sig_cs = of_string sig_bytes in
  let (r, s) = P256.Dsa.cstruct_to_sig sig_cs |> function Ok x -> x | Error _ -> failwith "invalid signature" in
  let digest = digest_string (Bytes.of_string message) |> to_raw_string |> of_string in
  P256.Dsa.verify ~key:signing_pub digest (r, s)

let encrypt message =
  let _, _, _, encrypting_pub = get_keys () in
  let msg_cs = of_string message in
  let ciphertext = Rsa.OAEP(SHA256).encrypt ~key:encrypting_pub msg_cs empty |> function Ok ct -> ct | Error _ -> failwith "encrypt failed" in
  encode_string (to_string ciphertext)

let decrypt ciphertext_b64 =
  let _, _, encrypting_priv, _ = get_keys () in
  let ct_bytes = decode_string ciphertext_b64 in
  let ct_cs = of_string ct_bytes in
  let plaintext = Rsa.OAEP(SHA256).decrypt ~key:encrypting_priv ct_cs empty |> function Some pt -> pt | None -> failwith "decrypt failed" in
  to_string plaintext

let () =
  init ();
  let sig_ = sign "Hello World" in
  Printf.printf "Signature: %s\n" sig_;
  Printf.printf "Verify: %b\n" (verify "Hello World" sig_);
  let enc = encrypt "Secret message" in
  Printf.printf "Encrypted: %s\n" enc;
  Printf.printf "Decrypted: %s\n" (decrypt enc)
